features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher-new/source/src/app/components/Apis/Create/GraphQL/APICreateGrpahQL.jsx

```diff
@@ -0,0 +1,369 @@
+            validUpdated.endpointConfig.empty = !api.endpointConfig;
+            // TODO we need to add the already existing error for (context)
+            // by doing an api call ( the swagger definition does not contain such api call)
+
+            return { api: changes, valid: validUpdated };
+        });
+    }
+
+    /**
+     * Make API POST call and create send GraphQL file or URL
+     * @memberof APICreateGraphQL
+     */
+    createGraphQLAPI() {
+        this.setState({ loading: true });
+        const newApi = new API();
+        const { graphQLBean, api } = this.state;
+        const {
+            name, version, context, policies, endpointConfig, gatewayEnvironments, implementationType, operations,
+        } = api;
+        const uploadMethod = 'file';
+        const apiAttributes = {
+            name,
+            version,
+            context,
+            endpointConfig,
+            gatewayEnvironments,
+            policies,
+            operations,
+        };
+        const apiData = {
+            additionalProperties: JSON.stringify(apiAttributes),
+            implementationType,
+            [uploadMethod]: graphQLBean[uploadMethod],
+            file: graphQLBean.file,
+        };
+
+        newApi
+            .importGraphQL(apiData)
+            .then((response) => {
+                Alert.success(`${name} API Created Successfully.`);
+                const uuid = response.obj.id;
+                const redirectURL = '/apis/' + uuid + '/overview';
+                this.setState({ loading: false });
+                this.props.history.push(redirectURL);
+            })
+            .catch((errorResponse) => {
+                this.setState({ loading: false });
+                console.error(errorResponse);
+                const error = errorResponse.response.obj;
+                const messageTxt = 'Error[' + error.code + ']: ' + error.description + ' | ' + error.message + '.';
+                Alert.error(messageTxt);
+            });
+    }
+    updateFileErrors(newValid) {
+        this.setState({ valid: newValid });
+    }
+
+    handleNext = () => {
+        const { activeStep, graphQLBean, valid } = this.state;
+        const { uploadMethod } = 'file';
+
+        const validNew = JSON.parse(JSON.stringify(valid));
+
+        // Handling next ( getting graphQL file info and validating)
+        if (activeStep === 0) {
+            if (isEmpty(graphQLBean)) {
+                if (uploadMethod === 'file') {
+                    validNew.graphQLFile.empty = true;
+                }
+                this.setState({ valid: validNew });
+                return;
+            }
+            // No errors so let's fill the inputs with the graphQLBean
+            if (graphQLBean.info) {
+                if (graphQLBean.info.operations) {
+                    this.setState(({ api }) => {
+                        const changes = api;
+                        changes.operations = graphQLBean.info.operations;
```

Comment:
seems like we are modifying the original state because we are mutating the original object. Its better if we can use the spread operator the copy the properties.

Reply:
Do you mean like this?
 this.setState((state) {
api.operations : [...state.api.operations,  graphQLBean.info.operations]
})

Reply:
This is not modifying the original object since it is defined at the sate

Reply:
it is modifying as js objects are pass by reference. after assigning like this
const changes = api;

even if you are only changing the **changes** variable you will still be modifying original object api which is in state.

you can find more details [here](http://dushansilva.com/2019/08/07/immutability-in-javascript/)

Reply:
Fixed